File: Python/programmers/kakao/level1/실패율.py
# sorted에 dictionary를 넘기면 dictionary의 keys가 들어간다.

def solution(N, stages):
    answer = {}
    L = len(stages)
    for i in range(1, N+1):
        if L != 0:
            c = stages.count(i)
            answer[i] = c/L
            L -= c
        else:
            answer[i] = 0
        
    return sorted(answer, key=lambda x: answer[x], reverse=True)


N = 5
stages = [2, 1, 2, 6, 2, 4, 3, 3]
print(solution(N, stages)) # [3,4,2,1,5]

N = 4
stages = [4,4,4,4,4]
print(solution(N, stages)) # [4,1,2,3]

File: Java/programmers/kakao/level1/실패율.java
package CodeTest.Java.programmers.kakao.level1;

import java.util.*;

public class 실패율 {
    public List<Integer> solution(int N, int[] stages) {        
        int[] Fail = new int[N+2];
        int[] Total = new int[N+1];
        
        for (int stage : stages) {
            Fail[stage]++;
        }
        
        Total[N] = Fail[N]+Fail[N+1];
        for (int i=N-1; i>=1; i--) {
            Total[i] = Fail[i]+Total[i+1];
        }

        HashMap<Integer, Double> failMap = new HashMap<>();
        for (int i=1; i<=N; i++) {
            if (Fail[i] == 0 || Total[i] == 0) {
                failMap.put(i, 0.0);
            } else {
                failMap.put(i, (double) Fail[i]/Total[i]);
            }
        }
             
        List<Integer> answer = new ArrayList<>(failMap.keySet());
        Collections.sort(answer, (o1, o2) -> Double.compare(failMap.get(o2), failMap.get(o1)));
        return answer;
    }
}
